docs/namespace_config_spec.md

```diff
@@ -0,0 +1,281 @@
+`Namespace` or `Context`), the user can get context help from her IDE while
+writing the configuration.
+
+## Notation
+
+The syntax is specified using Extended Backus-Naur Form (EBNF):
+
+```ebnf
+Production  = production_name "=" [ Expression ] "." .
+Expression  = Alternative { "|" Alternative } .
+Alternative = Term { Term } .
+Term        = production_name | token [ "…" token ] | Group | Option | Repetition .
+Group       = "(" Expression ")" .
+Option      = "[" Expression "]" .
+Repetition  = "{" Expression "}" .
+```
+
+Productions are expressions constructed from terms and the following operators, in increasing precedence:
+
+```ebnf
+|   alternation
+()  grouping
+[]  option (0 or 1 times)
+{}  repetition (0 to n times)
+```
+
+Lower-case production names are used to identify lexical tokens. Non-terminals are in CamelCase. Lexical tokens are enclosed in double quotes "" or back quotes ``.
+
+The form `a … b` represents the set of characters from a through b as alternatives. The horizontal ellipsis `…` is also used elsewhere in the spec to informally denote various enumerations or code snippets that are not further specified.
+
+## Configuraton text representation
+
+The configuration is encoded in UTF-8.
+
+## Lexical elements
+
+### Comments
+
+1. Line comments start with the character sequence // and stop at the end of the line.
+2. General comments start with the character sequence /* and stop with the first subsequent character sequence */.
+
+### Identifiers
+
+Identifiers name program entities such as variables and types. An identifier is a sequence of one or more letters and digits. The first character in an identifier must be a letter.
+
+```ebnf
+identifier = letter { letter | digit } .
+digit      = "0" … "9"
+letter     = "A" … "Z" | "a" … "z" | "-" | "_"
+```
+
+### Keywords
+
+The configuration language has the following keywords:
+
+* `class`
+* `implements`
+* `metadata`
+* `relations`
+* `permissions`
+* `this`
+* `ctx`
+* `id`
+
+### Builtin Types
+
+The following types are built in:
+
+* `Context`
+* `Namespace`
+* `Array<Namespace>` or `Namespace[]`
+* boolean
+
+In TypeScript, they would be defined as follows:
+
+```typescript
+type Context = { subject: never }
+
+interface Namespace {
+    metadata: { id: number }
+    relations?: { [relation: string]: Namespace[] }
+    permissions?: { [method: string]: (ctx: Context) => boolean }
+}
+
+interface Array<Namespace> {
+  includes(searchElement: Namespace): boolean;
+  some(callbackfn: (value: Namespace) => boolean): boolean;
+}
+```
+
+### Operators
+
+The following character sequences represent boolean operators:
+
+| Operator | Signature | Semantic |
+| --- | --- | --- |
+| `&&` | *x* `&&` *y* | true iff. *x* and *y* are true |
+| `\|\|` | *x* `\|\|` *y* | true iff. *x* or *y* are true |
+| `==` | *x* `==` *y* | true iff. *x* equals *y* |
+| `!=` | *x* `!=` *y* | ture iff. *x* does not equal *y* |
```

Comment:
I think kevgo has a point here. Novice JS users will probably not care about this, but experienced devs will for sure ask themselves what this means and might also be confused as to why `===` is not working. We should also explain what `equal` means here?

Reply:
I agree that `===` is clunky and the whole situation is far from ideal. Here are a few things that I think we should keep in mind. 

The `==` operator has [very well defined behavior](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Equality) in JS. If we use `==` in our language, we have to remove the parts from its spec where it tries to convert the operands into types that are comparable. Technically, this is within our mission of definining a "subset" of TypeScript. 

I am concerned that removing elements from the spec of `==` effectively changes its behavior. This will confuse humans who likely expect a subset where the parts left in the language are "normal" JS. The TS tooling we plan on using (type checker, linter, unit tests running in Node) will always assume the JS behavior. I want to avoid situations where the behavior of a Keto spec is different in tests than when running inside Keto. So I guess I propose that we at least test this before finalizing this spec.

For reference, the `===` operator, while uglier, [already does what we want](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality). Unless "equality" means something completely different in Keto than it does in JS. Is that the case here?

Reply:
OK this makes sense. Then we will *only* support the `===` and `!==` operators, as this captures best the semantics of the rules in intersection with the semantics of JS.

Reply:
On second thought, since we right now in the spec only allow relations with type `T[]`, checks for equality are actually done implicitly via `includes`. This means we can take direct equality out of the spec for now.

Reply:
> OK this makes sense. Then we will _only_ support the `===` and `!==` operators, as this captures best the semantics of the rules in intersection with the semantics of JS.

I agree with this

> On second thought, since we right now in the spec only allow relations with type `T[]`, checks for equality are actually done implicitly via `includes`. This means we can take direct equality out of the spec for now.

True, none of the examples actually use an equality operator. Are there use cases where this is needed actually?